### PR TITLE
Right sidebar, conf top, enable/disable shadow

### PR DESCRIPTION
**Pull request by ababol**
I have added a few features/configuration to the sidebar.

- `pullRight` : Place the sidebar on the right
- `topSidebar` : Sidebar distance from the top
-  `enableShadow` : Enable sidebar shadow

Fix https://github.com/balloob/react-sidebar/issues/7

Demo : http://meerkat.ovh

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,6 +35,9 @@ Supported props
 | touch | boolean | true | If touch gestures should be enabled |
 | touchHandleWidth | number | 20 | Width in pixels you can start dragging from the edge when the sidebar is closed. |
 | dragToggleDistance | number | 30 | Distance the sidebar has to be dragged before it will open/close after it is released. |
+| pullRight | boolean | false | Place the sidebar on the right |
+| topSidebar | number | 0 | Sidebar distance from the top. |
```
Comment (balloob) [inline README.md]: React sidebar positions itself using a `div` that uses `position: absolute` to have it take over the whole area of the `div`. If you wrap React sidebar in another `div` that sets the `position` style attribute, it will not take over the whole screen but only the parent `div`. Therefore, I can't accept the `topSidebar` feature.

File: src/sidebar.js [hunk 2]
```diff
@@ -209,6 +202,36 @@ class Sidebar extends React.Component {
 
     let rootProps = {
       style: styles.root,
+    };
+
+    rootProps.style.top = this.props.topSidebar;
+
+    if (this.props.pullRight) {
+      if (this.props.enableShadow) {
+        sidebarStyle.boxShadow = '-2px 2px 4px rgba(0, 0, 0, 0.15)';
```
Comment (balloob) [inline src/sidebar.js]: Please do not modify the `sidebarStyle` style property directly but only create new ones using the `update` method (as is done 2 lines below). Probably better to merge the two calls to `update` in one:

```javascript
const sidebarStyleToMerge = {
  right: 0,
  transform: 'translateX(100%)',
  WebkitTransform: 'translateX(100%)',
}

if (this.props.shadow) {
  sidebarStyleToMerge.boxShadow = '-2px 2px 4px rgba(0, 0, 0, 0.15)';
}

sidebarStyle = update(sidebarStyle, { $merge: sidebarStyleToMerge });
```

File: src/sidebar.js [hunk 3]
```diff
@@ -343,6 +383,9 @@ Sidebar.defaultProps = {
   transitions: true,
   touch: true,
   touchHandleWidth: 20,
+  topSidebar: 0,
+  pullRight: false,
+  enableShadow: true,
```
Comment (balloob) [inline src/sidebar.js]: Could we rename this to `shadow` ? The boolean value implies the enable part of the name.

File: src/sidebar.js [hunk 4]
```diff
@@ -35,7 +30,6 @@ const styles = {
     right: 0,
     bottom: 0,
     overflow: 'auto',
-    transition: 'left .3s ease-out',
```
Comment (balloob) [inline src/sidebar.js]: Instead of setting this attribute each render we can simply enable the transition for both `left` and `right` attribute by separating them with a comma:
```css
transition: left .3s ease-out, right .3s ease-out
```

File: src/sidebar.js [hunk 5]
```diff
@@ -209,6 +202,36 @@ class Sidebar extends React.Component {
...
+      }
+      sidebarStyle = update(sidebarStyle, { $merge: {
+        right: 0,
+        transform: 'translateX(100%)',
+        WebkitTransform: 'translateX(100%)',
+      }});
+
+      contentStyle.transition = 'right .3s ease-out';
+
+      styles.dragHandle.left = 0;
```
Comment (balloob) [inline src/sidebar.js]: Please do not mutate constants in the top of the file but use the `update` method to create a new object.

**Comment by balloob** [comment]
Thanks for the PR. I've just added a few comments.

Tl;dr:

 * `topSidebar` property can be achieved by wrapping React sidebar in a styled `div` so no need to add it to React sidebar.
 * Avoid mutating the style constants at the top of the file but create new objects using the update method instead ([update docs can be found here](https://facebook.github.io/react/docs/update.html)).

File: src/sidebar.js (hunk 2 above)
Comment (ababol) [inline src/sidebar.js]: :+1:

File: src/sidebar.js (hunk 3 above)
Comment (ababol) [inline src/sidebar.js]: :+1:

File: src/sidebar.js (hunk 4 above)
Comment (ababol) [inline src/sidebar.js]: :+1:

File: README.md (hunk 1 above)
Comment (ababol) [inline README.md]: Would you accept it if I add :warning:  *Use with caution* :warning:  in the README :smile: ?!

This feature is useful on my project and it is working like a charm...

**Comment by ababol** [comment]
Hi @balloob, thanks for the comments.

I have made some changes, let me know if I should rebase before your merge.

File: src/sidebar.js [hunk 6]
```diff
@@ -94,13 +89,18 @@ class Sidebar extends React.Component {
     // filter out if a user starts swiping with a second finger
     if (!this.isTouching()) {
       let touch = ev.targetTouches[0];
+      let tcX = touch.clientX;
+      if (this.props.pullRight) {
+        tcX = this.state.sidebarWidth - tcX;
```
Comment (balloob) [inline src/sidebar.js]: We shouldn't couple the fact that the sidebar is in pull right mode with the code that handles the touches. The logic should be incorporated in the render method where we're actually deciding on the props and are pull right aware.

Comment (ababol) [inline src/sidebar.js]: Oh yeah to be honest I knew that but I was lazy, I'm sorry :/.

If I am correct we should edit the [touchSidebarWith](https://github.com/ababol/react-sidebar/blob/right-sidebar/src/sidebar.js#L192) method and do the logic here ?

**Comment by boristian** [comment]
Hello,
will this be merged? I would love to have that features.

**Comment by balloob** [comment]
The features are not done yet so can't merge.

File: src/sidebar.js (hunk 6 above)
Comment (ababol) [inline src/sidebar.js]: :up:  :surfer:

Comment (balloob) [inline src/sidebar.js]: Sorry, I meant the methods `inCancelDistanceOnScroll()` and `touchSidebarWidth()`. They are actually using the coordinates and should be made pullRight aware.

**Comment by ababol** [comment]
Hey @balloob , sorry about the delay!

I have made the changes :surfer:, can you take a look?

**Comment by balloob** [comment]
Sorry for the delay, looks good!

**Comment by ababol** [comment]
Awesome! Thanks :+1:
